### ISSUE TITLE
Windows: udp receives no data, maybe a SetFileCompletionNotificationModes issue.

### ISSUE THREAD

**Issue by BillHoo** [libuv/libuv#1672]
* **Version**: 1.11.0
* **Platform**: Windows7 SP1 x64

Hi uv team,

I have encounter a bug on two PCs, one is my test computer, the other is on customer side(so that one has no tech info).

[Minimum demo]
Use offical unit test or DHCP demo from official doc is enough.

[What happend]
When I run unittest or my own test case on that computer, udp send is success, I can grab the sending datagram and server response from WireShark, but the response is not transfered to my application, i.e. uv_udp_recv_cb is nerver called.

Besides, all the libuv API's returns 0, told me that there's no error, but things goes wrong indead.
When I get into the code, I found this:

uv_udp_set_socket:

    /* All known Windows that support SetFileCompletionNotificationModes */
    /* have a bug that makes it impossible to use this function in */
    /* conjunction with datagram sockets. We can work around that but only */
    /* if the user is using the default UDP driver (AFD) and has no other */
    /* LSPs stacked on top. Here we check whether that is the case. */

and the "info.ProtocolChain.ChainLen" equals 2 on that computer(so it seems the hack described above is not performed on that computer, maybe this is the clue), and the rest of my testing computers get the value of 1.

[Expected]
libuv's udp component is fine on all kinds of windows systems or can go back to normal mode if the IOCP hack failed.

[Question]
1. Is there a way to check whether the system use the default UDP driver (AFD) or not?
2. Is there a way to check there's other LSP stack on top of UDP?
3. If neighter of them can be done, is there a way to set libuv back to normal mode instead of IOCP on UDP side?
4. If 1-3 can't be done, is there a way to return a error from any of the API so I can do something at least.
5. If all of them can't be done, what can I do on this senario?

Thanks in advance,
Bill.

**Comment by bnoordhuis** [comment]
To questions 1-2: You can check for LSPs with Windows Defender's Software Explorer tool.

To questions 3-5: You could patch libuv to always take the path you want.  We're open to patches that improve feature detection.

**Comment by BillHoo** [comment]
Thanks,

For 1-2, I've found a command:
`netsh winsock show catalog`
to check if there's more than one LSP, result is that problematic computer have two.

But for 3-5, I'm not a Windows Network Developing expert, so I'm expecting some official API or official advise to handle this senario.

There's already an official comments in the code says it's a bug from Microsoft, but what can we do when we really meet this bug? Because if we do meet this senario in some of our customer's computer, it seems no way to avoid it? That means we can not use libuv to handle UDP stuffs anymore, what a pity :(

**Comment by bnoordhuis** [comment]
> so I'm expecting some official API or official advise to handle this senario

Let me unpack what I wrote.  Libuv already tries to detect and route around that scenario but one or either option below may apply:

1. Libuv's auto-detection is defective.  Not exactly impossible but not exactly plausible either as it has worked well for years.
2. One of the LSPs is defective, breaking auto-detection.  Microsoft deprecated LSPs for exactly that reason: most are extremely buggy and some are outright malware.

Here is a simple patch you can try that will avoid the problematic code path.  Please try it and let me know what happens.
```diff
diff --git a/src/win/winapi.c b/src/win/winapi.c
index 4ccdf0a5f..c108dc769 100644
--- a/src/win/winapi.c
+++ b/src/win/winapi.c
@@ -123,9 +123,6 @@ void uv_winapi_init(void) {
       kernel32_module,
       "GetQueuedCompletionStatusEx");
 
-  pSetFileCompletionNotificationModes = (sSetFileCompletionNotificationModes)
-    GetProcAddress(kernel32_module, "SetFileCompletionNotificationModes");
-
   pCreateSymbolicLinkW = (sCreateSymbolicLinkW)
     GetProcAddress(kernel32_module, "CreateSymbolicLinkW");
 
```

**Comment by BillHoo** [comment]
Got it, thank you for the details reply. I'll try it and let you know.

BTW, is the same thing happens to TCP side?

**Comment by bnoordhuis** [comment]
Yep.  The patch will make it skip the `SetFileCompletionNotificationModes()` call there as well (although it may not need to.)

**Comment by BillHoo** [comment]
I can't reproduce the problem on that testing computer again.

Before I got your patch, I found there was a malware by running
`netsh winsock show catalog`
so I uninstall it to verify if its the point. And after uninstall, everything goes well.

After that, I reinstall that malware many times, but still can't reproduce the problem :(

Since the patch can't be tested, I have to confirm something about it:

1. When `pSetFileCompletionNotificationModes` is set to NULL, can libuv still running well on both UDP and TCP side?

2. I noticed that the `pSetFileCompletionNotificationModes` is aimed to enhance the performance, so does it slow down libuv very much without it?

Thanks.

**Comment by bzoz** [comment]
1) It is NULL in Windows XP. It should work, but no one is testing that so there is no guarantee since we do not support XP anymore.

2) There is a [comment in winsock.c](https://github.com/libuv/libuv/blob/fd39cec4b848b27be6bf3f8a3e0bce29474366bb/src/win/winsock.c#L277):
"However it is desirable to use SetFileCompletionNotificationModes because it yields a massive performance increase."

BTW, there is a PR for v2 to remove LSP support: https://github.com/libuv/libuv/pull/1066
